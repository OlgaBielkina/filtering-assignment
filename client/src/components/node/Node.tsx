import React from 'react';
import { Node as NodeType } from '../../types/types';
import { Accordion } from '../accordion/Accordion';
import { AccordionContent } from '../accordion/AccordionContent';
import { AccordionToogle } from '../accordion/AccordionToogle';
import { Checkbox } from '../checkbox/Checkbox';
import styles from './Node.module.scss';

export interface Props {
    node: NodeType;
    onSelect: (nodeId: string, selected: boolean) => void;
}

export const Node: React.FC<Props> = ({ node, onSelect }) => {
    const changeHandler = () => {
        onSelect(node.id, !node.selected);
    };

    return (
        <Accordion>
            <AccordionToogle collapsible={!!node.children.length}>
                <Checkbox label={node.name} onChange={changeHandler} checked={node.selected} className={styles.row} />
            </AccordionToogle>
            <AccordionContent>
                <div className={styles.offset}>
                    {node.children.map((child) => (
                        <Node key={child.id} node={child} onSelect={onSelect} />
                    ))}
                </div>
            </AccordionContent>
        </Accordion>
    );
};
