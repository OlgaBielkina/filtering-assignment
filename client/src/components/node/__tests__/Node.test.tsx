import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { Node, Props } from '../Node';

describe('Node', () => {
    const props: Props = {
        node: {
            id: '1',
            count: 1,
            parent: '0',
            name: 'test name 1',
            children: [
                {
                    id: '2',
                    count: 1,
                    parent: '1',
                    name: 'test name 2',
                    children: [],
                    selected: false,
                },
                {
                    id: '3',
                    count: 1,
                    parent: '1',
                    name: 'test name 3',
                    children: [
                        {
                            id: '4',
                            count: 1,
                            parent: '3',
                            name: 'test name 4',
                            children: [],
                            selected: false,
                        },
                    ],
                    selected: false,
                },
            ],
            selected: false,
        },
        onSelect: jest.fn(),
    };

    it('should render correctly', () => {
        const { asFragment } = render(<Node {...props} />);

        expect(asFragment()).toMatchSnapshot();
    });

    it('should render all nodes collapsed', () => {
        const { getAllByTestId } = render(<Node {...props} />);

        getAllByTestId('accordionContent').forEach((content) => {
            expect(content).toHaveClass('hidden');
        });
    });

    it('should call onSelect when user pressed on checkbox', () => {
        const { getByLabelText } = render(<Node {...props} />);

        fireEvent.click(getByLabelText(props.node.name));

        expect(props.onSelect).toBeCalledWith(props.node.id, !props.node.selected);
    });

    it('should expand node subtree when user pressed expan button', () => {
        const { getAllByText, getAllByTestId } = render(<Node {...props} />);

        fireEvent.click(getAllByText('expand_more')[0]);

        expect(getAllByTestId('accordionContent')[0]).not.toHaveClass('hidden');
    });
});
