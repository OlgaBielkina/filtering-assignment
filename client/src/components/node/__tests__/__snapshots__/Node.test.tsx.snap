// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Node should render correctly 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <label
      class="root row"
    >
      <input
        class="input"
        type="checkbox"
      />
      <span
        class="label"
      >
        test name 1
      </span>
    </label>
    <span
      class="material-icons icon"
    >
      expand_more
    </span>
  </div>
  <div
    class="hidden"
    data-testid="accordionContent"
  >
    <div
      class="offset"
    >
      <div
        class="root"
      >
        <label
          class="root row"
        >
          <input
            class="input"
            type="checkbox"
          />
          <span
            class="label"
          >
            test name 2
          </span>
        </label>
      </div>
      <div
        class="hidden"
        data-testid="accordionContent"
      >
        <div
          class="offset"
        />
      </div>
      <div
        class="root"
      >
        <label
          class="root row"
        >
          <input
            class="input"
            type="checkbox"
          />
          <span
            class="label"
          >
            test name 3
          </span>
        </label>
        <span
          class="material-icons icon"
        >
          expand_more
        </span>
      </div>
      <div
        class="hidden"
        data-testid="accordionContent"
      >
        <div
          class="offset"
        >
          <div
            class="root"
          >
            <label
              class="root row"
            >
              <input
                class="input"
                type="checkbox"
              />
              <span
                class="label"
              >
                test name 4
              </span>
            </label>
          </div>
          <div
            class="hidden"
            data-testid="accordionContent"
          >
            <div
              class="offset"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
